frontend/components/CellOutput.js

```diff
@@ -683,3 +695,44 @@ export let highlight = (code_element, language) => {
         }
     }
 }
+
+/**
+ * Generates a copy button for Markdown code blocks and copies them into the clipboard.
+ * @param {HTMLElement | null} pre - The <pre> element containing the code block.
+ */
+export const generateCopyCodeButton = (pre) => {
+    if (!pre) {
+        console.error('Error: pre is null.');
+        return;
+    }
+
+    const copyToClipboard = () => {
+        const range = document.createRange()
+        range.selectNode(pre)
+        const txt = range.startContainer.textContent || ''
+        navigator.clipboard
+            .writeText(txt)
+            .then(() => {
+                console.log("Code copied to clipboard:\n" + txt)
+            })
+            .catch((error) => {
+                console.error("Error copying code to clipboard:", error)
+            })
+    }
```

Comment:
I would move the copy function outside this function, to avoid creating a closure for every cell output.

check https://developer.mozilla.org/en-US/docs/Web/API/Element/closest

```suggestion
/** Copies the contents of an HTML Element into the clipboard */
const copyToClipboard = (e) => {
  const pre = e.target.parentNode.parentNode // or something similar, to find the `pre`. See `.closest()`
  const range = document.createRange()
  range.selectNode(pre)
  const txt = range.startContainer.textContent || ''
  navigator.clipboard
    .writeText(txt)
    .then(() => {
      console.log("Code copied to clipboard:\n" + txt)
    })
    .catch((error) => {
       console.error("Error copying code to clipboard:", error)
    })
}

/**
 * Generates a copy button for Markdown code blocks and copies them into the clipboard.
 * @param {HTMLElement | null} pre - The <pre> element containing the code block.
 */
export const generateCopyCodeButton = (pre) => {
    if (!pre) {
        console.error('Error: pre is null.');
        return;
    }
```